### PR TITLE
Add ddpm kandinsky

### PR DESCRIPTION
**Pull request by yiyixuxu**
This PR will refactor Kandinsky pipelines to work with both ddpm and ddim scheduler 

@sayakpaul @patrickvonplaten 
currently the scheduler config in Kandinsky repo is for ddim, is there anyway we can also save ddpm scheduler config some where? instead having to manually create the ddpm config when user wants swap it the scheduler? 

```python
from diffusers import KandinskyPipeline, KandinskyPriorPipeline, DDPMScheduler

import torch
import numpy as np

device = "cuda"

# # inputs
prompt= "red cat, 4k photo"

# # create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained("kandinsky-community/kandinsky-2-1-prior")
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device=device).manual_seed(0)
out = pipe_prior(prompt, num_inference_steps=5, generator=generator, )
image_emb = out.image_embeds
zero_image_emb = out.negative_image_embeds


# create diffuser pipeline
model_dtype = torch.float16
pipe = KandinskyPipeline.from_pretrained("kandinsky-community/kandinsky-2-1", torch_dtype =model_dtype )
pipe.to(device)

ddpm_scheduler = DDPMScheduler(
  clip_sample= True,
  clip_sample_range=2.0,
  sample_max_value=None,
  num_train_timesteps= 1000,
  prediction_type= "epsilon",
  variance_type= "learned_range",
  thresholding= True,
  beta_schedule= "linear",
  beta_start= 0.00085,
  beta_end=0.012
)
pipe.scheduler = ddpm_scheduler

generator = torch.Generator(device="cuda").manual_seed(0)
out = pipe(
    prompt,
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=100,
    generator=generator )

image = out.images[0]
```
yiyi_test_pipe_ddpm3_out

### CODE REVIEW

**Comment by yiyixuxu** [comment]
still not able to reproduce the doc examples since I think the initial noise construction is a little bit different now with the most recent UI change for negative prompts but like these new images too 

image
image

**Comment by yiyixuxu** [comment]
@sayakpaul yeah but then user has to know the new args for scheduler, which is pretty much everything - the config for ddim and ddpm is really different for kandinsky

ddim

```python
{
   "num_train_timesteps": 1000,
   "beta_schedule":  "linear",
   "beta_start": 0.00085,
   "beta_end":0.012,
    "clip_sample" : false,
    "set_alpha_to_one" : false,
    "steps_offset" : 1,
    "prediction_type" : "epsilon",
    "thresholding" : false
}
```
ddpm 
```python
DDPMScheduler(
  clip_sample= True,
  clip_sample_range=2.0,
  sample_max_value=None,
  num_train_timesteps= 1000,
  prediction_type= "epsilon",
  variance_type= "learned_range",
  thresholding= True,
  beta_schedule= "linear",
  beta_start= 0.00085,
  beta_end=0.012
)
```

File: src/diffusers/schedulers/scheduling_ddpm.py [hunk 1]
```diff
@@ -367,9 +367,14 @@ def step(
             )
 
         # 3. Clip or threshold "predicted x_0"
+        if self.config.sample_max_value is None and self.config.clip_sample:
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddpm.py:370]: I don't really understand this change - what are we doing here exactly? Is the problem that thresholding is done after clamping in Kandinsky?

File: src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py [hunk 2]
```diff
@@ -390,8 +408,13 @@ def __call__(
         image_embeds = torch.cat([negative_image_embeds, image_embeds], dim=0).to(
             dtype=prompt_embeds.dtype, device=device
         )
+        if hasattr(self.scheduler, "custom_timesteps"):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:411]: Hmm I don't think this logic makes much sense thb. If we need a different spacing for Kandinsky, let's add this to the DDPMScheduler similarly to: https://github.com/huggingface/diffusers/blob/77f9137f103b04f137b6c487814ea58599ebc200/src/diffusers/schedulers/scheduling_ddim.py#L306

**Comment by patrickvonplaten** [comment]
Also the design of the this PR needs to be adapted tough. I think this PR introduces complex/weird design choices

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: @patrickvonplaten 

I just need to clamp it and then thresholding 

Currently, it will only do thresholding __OR__ clamping, never both together, so if you set `clip_smple` to be `True`, it will NOT clip it if `threholding` is also `True`.  I checked the config for IF,  both`clip_sample` and `thresholding` are set to be `True` there, but only threholding will be applied https://huggingface.co/DeepFloyd/IF-I-XL-v1.0/blob/main/scheduler/scheduler_config.json

I'm not sure why we need this `if ... else... `  logic , I think it's confusing. if we only want threholding, we should just set `clip_sample` to be `False`. Removing it now will break IF I think

Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: The reason for why thresholding and clipping is done in an `if, else` statement is because it's more or less the same operation.

If you do thresholding your output **has to be** clipped to [-1, 1] as can be seen here: https://github.com/huggingface/diffusers/blob/59aefe9ea6d56245d257439f757384fb1ac5d477/src/diffusers/schedulers/scheduling_ddim.py#L280

So in the case of IF even if the `self.config.clip_sample` would be run afterward it would make no difference as the output is already scaled to be between -1 and 1. We could/should probably add a warning that running `thresholding` and `clipping` together doesn't make sense and should be avoided. For IF it would probably be best if we just set `clip_sample=False` or is this used for anything else ? cc @williamberman 

@yiyixuxu for Kandinsky the current "working" scheduler config is:
```
DDPMScheduler(
  clip_sample= True,
  clip_sample_range=2.0,
  sample_max_value=None,
  num_train_timesteps= 1000,
  prediction_type= "epsilon",
  variance_type= "learned_range",
  thresholding= True,
  beta_schedule= "linear",
  beta_start= 0.00085,
  beta_end=0.012
)
```

with this PR. I'm fairly certain that the following config **without** any changes to `scheduling_ddim.py` is the same:

```
DDPMScheduler(
  clip_sample= False,
  sample_max_value=2.0,
  num_train_timesteps= 1000,
  prediction_type= "epsilon",
  variance_type= "learned_range",
  thresholding= True,
  beta_schedule= "linear",
  beta_start= 0.00085,
  beta_end=0.012
)
```

Could you try this out?

Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: I think the current logic since clip sample is the default, we just prioritize thresholding if its also set. So we're ok with both being set but if so we just do the threshold clipping

Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: And yeah it's an if/else because thresholding is a form of clipping so it doesn't make sense to do both

Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: Guess overall it just doesn't make too much sense to set both no? Think a warning like "Your clip sample is useless and doesn't do anything" could make sense

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: > Could you try this out?

will try and report back :)

Comment (williamberman) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: yeah no issues with either

**Comment by patrickvonplaten** [comment]
Let's try to get this one in soon @yiyixuxu

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_ddpm.py:370]: @patrickvonplaten the cat image looks same to me! will revert 

is math the same, or just that difference too small that we can ignore? I understood clip sample and dynamic thresholding are trying to do the same thing, but when I looked at the formula, I thought clamping first would cause the threshold to be different and cause a difference in the result, no the case it seems? 
yiyi_test_pipe_ddpm_patrick_out

File: src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py (hunk 2 above)
Comment (yiyixuxu) [reply src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py]: @patrickvonplaten 
looked into this, I think kandinsky's method of calculating step ratio is just **slightly** different from what's already in DDPM, maybe we don't need to make any changes as long as it does not affect generation quality? 

cc @williamberman because Kandinsky timesteps is is exactly same as unclip 

- kandinsky/ unclip scheduler
```python
step_ratio = (self.scheduler.config.num_train_timesteps - 1) / (num_inference_steps - 1)
timesteps = (np.arange(0, num_inference_steps) * step_ratio).round()[::-1].copy().astype(np.int64)
```

- ddpm 
```
step_ratio = self.config.num_train_timesteps // self.num_inference_steps
timesteps = (np.arange(0, num_inference_steps) * step_ratio).round()[::-1].copy().astype(np.int64)
```
compared the cheeseburger monster, the one above is generated with Kandinsky timesteps; the one below is generated with ddpm scheduler without any change. I can see the eyes and mouth are a little bit different for the monster on the right but it really doesn't affect quality

yiyi_test_pipe_ddpm_burger_out_original

yiyi_test_pipe_ddpm_burger_out

only super minor difference here too

yiyi_test_pipe_ddpm_hair_out_original

    
yiyi_test_pipe_ddpm_hair_out

**Comment by yiyixuxu** [comment]
@patrickvonplaten ready for another review

File: src/diffusers/schedulers/scheduling_ddpm.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_ddpm.py]: Think math is the same here actually

**Review by patrickvonplaten — APPROVED** [review]
Cool that works!
